cou/steps/plan.py

```diff
@@ -42,10 +45,46 @@ async def generate_plan(analysis_result: Analysis) -> UpgradeStep:
         UpgradeStep(description="backup mysql databases", parallel=False, function=backup)
     )
 
-    upgrade_plan = UpgradeStep(
-        description="Application(s) upgrade plan", parallel=False, function=None
+    principle_upgrade_plan = await create_upgrade_group(
+        analysis_result=analysis_result,
+        description="Principal(s) upgrade plan",
+        target=target,
+        filter_function=lambda app: not isinstance(app, OpenStackSubordinateApplication),
     )
-    for app in analysis_result.apps:
+    plan.add_step(principle_upgrade_plan)
+
+    subordinate_upgrade_plan = await create_upgrade_group(
+        analysis_result=analysis_result,
+        description="Subordinate(s) upgrade plan",
+        target=target,
+        filter_function=lambda app: isinstance(app, OpenStackSubordinateApplication),
+    )
+    plan.add_step(subordinate_upgrade_plan)
+
+    return plan
+
+
+async def create_upgrade_group(
+    analysis_result: Analysis,
+    target: str,
+    description: str,
+    filter_function: Callable[[OpenStackApplication], bool],
+) -> UpgradeStep:
+    """Create upgrade group.
+
+    :param analysis_result: Result of the analysis.
+    :type analysis_result: Analysis
+    :param target: Target OpenStack version.
+    :type target: str
+    :param description: Description of the upgrade step.
+    :type description: str
+    :param filter_function: Function to filter applications.
+    :type filter_function: Callable[[OpenStackApplication], bool]
+    :return: Upgrade group.
+    :rtype: UpgradeStep
+    """
+    group_upgrade_plan = UpgradeStep(description=description, parallel=False, function=None)
+    for app in (app for app in analysis_result.apps if filter_function(app)):
```

Comment:
you called it filter_function, you might as well pass it to a filter ;)
```suggestion
    for app in filter(filter_function, analysis_result.apps):
```

Reply:
ok